NYTimes/video-captions-api
# validate uploaded vtt files

Currently when you upload a captions file, there is no validation of the uploaded content, so we have run into situations where a user manually uploads an invalid VTT file. This then causes downstream issues when that captions file is handed off to Amara for review.

This pull requests adds VTT validation to the upload provider. Currently only VTT is supported by checking the extension of the filename and running a validator. Longer term we should add support for other formats.

Also because there wasn't a great and well adopted VTT validator in the golang community, I created one based off the [VTT spec](https://www.w3.org/TR/webvtt1) and inspired by the mozilla's [vtt.js](github.com/mozilla/vtt.js) project. Longer term, the VTT validator should be broken out into it's own library, especially as we support other formats.

To test manually:

1. Start the server: `make run`
2. In a new terminal, set a vtt file: `vtt_file=vtt/testdata/sample.vtt`
3. Call the new captions endpoint:
```
cat $vtt_file | base64 | sed "s/\(.*\)/{\"provider\":\"upload\",\"caption_file\":{\"file\":\"\1\",\"name\":\"$(basename $vtt_file)\"}}/" | curl -H "Content-Type: application/json" -X POST -d @- http://localhost:8000/captions | jq
```
4. You should get a payload back that status is `delivered`
5. Now set one of the bad vtt files `vtt_file=vtt/testdata/cue-invalid-timestamp.vtt`
6. Run step 3
7. You should get an error message with a specific issue and line number:
```
{
  "error": "Error dispatching Job: [cue] invalid start timestamp, expecting: \"00:00:00.000\", got: \"00:11.00\" [line 3]"
}
``
